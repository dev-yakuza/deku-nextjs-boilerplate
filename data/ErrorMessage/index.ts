import { atom } from 'recoil'

export const errorMessageState = atom<string | undefined>({
  key: 'ERROR_MESSAGE_STATE',
  default: undefined,
})
