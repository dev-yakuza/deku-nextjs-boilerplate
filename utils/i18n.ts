import i18n from 'i18next'
import { initReactI18next } from 'react-i18next'
import { i18n as i18nConfig } from '../next-i18next.config'

i18n.use(initReactI18next).init({
  fallbackLng: 'en',
  resources: Object.assign(
    {},
    ...i18nConfig.locales.map((locale) => ({
      [locale]: {
        common: require(`../public/locales/${locale}/common.json`),
      },
    })),
  ),
})

export default i18n
