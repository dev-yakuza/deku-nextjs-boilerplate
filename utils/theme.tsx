import { createTheme } from '@mui/material/styles'

export const theme = createTheme({
  palette: {
    primary: {
      main: '#ff8e88',
    },
  },
  typography: {
    h1: {
      fontSize: '2rem',
    },
  },
})
