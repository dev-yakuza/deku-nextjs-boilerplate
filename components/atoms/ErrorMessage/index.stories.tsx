import { RecoilRoot, useSetRecoilState } from 'recoil'

import { errorMessageState } from 'data/ErrorMessage'

import type { ComponentMeta, ComponentStory } from '@storybook/react'

import { ErrorMessage } from '.'

export default {
  title: 'Atoms/ErrorMessage',
  component: ErrorMessage,
} as ComponentMeta<typeof ErrorMessage>

interface TestComponentProps {
  readonly errorMessage: string
}
const TestComponent = ({ errorMessage }: TestComponentProps) => {
  const setErrorMessage = useSetRecoilState(errorMessageState)

  return (
    <>
      <ErrorMessage />
      <button onClick={() => setErrorMessage(errorMessage)}>Show</button>
    </>
  )
}

const Template: ComponentStory<typeof TestComponent> = (args) => (
  <RecoilRoot>
    <TestComponent {...args} />
  </RecoilRoot>
)

export const Default = Template.bind({})
Default.args = {
  errorMessage: 'This is a sample error message',
}
