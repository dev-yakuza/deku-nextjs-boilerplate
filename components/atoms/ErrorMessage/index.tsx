import { useEffect, useState } from 'react'

import { Alert, Snackbar } from '@mui/material'
import { useRecoilState } from 'recoil'

import { errorMessageState } from 'data/ErrorMessage'

export const ErrorMessage = () => {
  const [open, setOpen] = useState(false)
  const [errorMessage, setErrorMessage] = useRecoilState(errorMessageState)

  const handleClose = () => {
    setOpen(false)
    setTimeout(() => setErrorMessage(undefined), 100)
  }

  useEffect(() => {
    if (errorMessage !== undefined && errorMessage !== '') {
      setOpen(true)
    }
  }, [errorMessage])

  return (
    <Snackbar
      open={open}
      autoHideDuration={3000}
      onClose={handleClose}
      anchorOrigin={{ vertical: 'bottom', horizontal: 'center' }}
    >
      <Alert onClose={handleClose} severity="error" sx={{ width: '100%' }}>
        {errorMessage}
      </Alert>
    </Snackbar>
  )
}
