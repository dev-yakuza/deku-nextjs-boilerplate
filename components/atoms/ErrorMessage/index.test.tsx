import { act, fireEvent, render, screen } from '@testing-library/react'
import { RecoilRoot, useSetRecoilState } from 'recoil'

import { errorMessageState } from 'data/ErrorMessage'
import { mockSnackbar } from 'utils/test'

import { ErrorMessage } from '.'

jest.useFakeTimers()

describe('<ErrorMessage />', () => {
  it('Rendered well', async () => {
    const TestComponent = () => {
      const setErrorMessage = useSetRecoilState(errorMessageState)

      return (
        <>
          <ErrorMessage />
          <button
            onClick={() => setErrorMessage('This is a test error message')}
          >
            Show
          </button>
        </>
      )
    }

    const { container } = render(
      <RecoilRoot>
        <TestComponent />
      </RecoilRoot>,
    )

    let snackbar = mockSnackbar.mock.calls[0][0]
    expect(snackbar.open).toBe(false)
    expect(snackbar.autoHideDuration).toBe(3000)
    expect(snackbar.onClose.name).toBe('handleClose')
    expect(snackbar.anchorOrigin).toEqual({
      vertical: 'bottom',
      horizontal: 'center',
    })

    let alert = snackbar.children
    expect(alert.type.render.name).toBe('Alert')
    expect(alert.props.onClose.name).toBe('handleClose')
    expect(alert.props.severity).toBe('error')
    expect(alert.props.sx).toEqual({ width: '100%' })
    expect(alert.props.children).toBe(undefined)

    mockSnackbar.mockClear()
    fireEvent.click(screen.getByText('Show'))

    snackbar = mockSnackbar.mock.calls[1][0]
    expect(snackbar.open).toBe(true)
    alert = snackbar.children
    expect(alert.props.children).toBe('This is a test error message')

    mockSnackbar.mockClear()
    act(() => {
      jest.runOnlyPendingTimers()
    })

    snackbar = mockSnackbar.mock.calls[0][0]
    expect(snackbar.open).toBe(false)
    alert = snackbar.children
    expect(alert.props.children).toBe('This is a test error message')

    mockSnackbar.mockClear()
    act(() => {
      jest.runOnlyPendingTimers()
    })

    snackbar = mockSnackbar.mock.calls[0][0]
    expect(snackbar.open).toBe(false)
    alert = snackbar.children
    expect(alert.props.children).toBe(undefined)

    expect(container).toMatchSnapshot()
  })
})
