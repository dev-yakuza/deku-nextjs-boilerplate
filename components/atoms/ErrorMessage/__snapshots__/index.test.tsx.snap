// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorMessage /> Rendered well 1`] = `
<div>
  <button>
    Show
  </button>
</div>
`;
